### ISSUE TITLE
Group API docs by purpose instead of collection

### ISSUE THREAD

**Issue by justinbmeyer** [canjs/canjs#3404]
__tldr;__ Our packages should be grouped by purpose instead of by collection.  This proposes to reorganize the documentation and give new users a more simplified understanding of CanJS.

> This was discussed at a [contributors meeting (8:30)](https://youtu.be/95OX6-qii24?t=8m30s)

## The Problem

The API docs sidebar looks like:

![canjs_-_api_docs](https://user-images.githubusercontent.com/78602/30554734-2c302f6a-9c6b-11e7-8c38-ff826838d5e4.png)

There are two problems with the API docs:

- It's huge and intimidating to new users
- It's primarily organized around collection (quality and likelyhood of use) instead of purpose.

#### Intimidating new users

Listing 60 packages makes CanJS seem way too big.  Most of these packages they will never have to use, or use very rarely.

#### Organized by collections instead of purpose 

While being organized by collection is useful at focusing people's attention on what they will primarily use, it does little to inform people what a given package actually does.

## Solution 

I would like to group our packages by purpose while still retaining the ability to "focus" new users on the APIs that they need to know.  

The following is a rough idea of what the sidebar might look like to a new user:

<img width="374" alt="untitled" src="https://user-images.githubusercontent.com/78602/30555662-cdd15734-9c6d-11e7-8cf7-160f2446845a.png">

The `[C]`, `[I]`, `[E]`, and `[L]` buttons toggle if core, infrastructure, ecosystem, or legacy packages should be shown. If someone toggled the buttons, in this case the`[I]`  button, they would see the infrastructure packages, likely color-coded:

<img width="374" alt="untitled" src="https://user-images.githubusercontent.com/78602/30555699-dfe14470-9c6d-11e7-8bc6-fd3ff2bb5045.png">

The following shows most of CanJS's packages organized this way:

- Observables
  - can-compute __I__
  - can-define __C__
  - can-event __I__
  - can-observation __I__
  - can-simple-map __I__
  - can-stream __E__
  - can-stream-kefir __E__
  - can-define-stream __E__
  - can-define-stream-kefir __E__

- Views
  - can-component __C__
  - can-stache __C__
  - can-stache-bindings __C__
  - can-view-callbacks __I__
  - can-view-live __I__
  - can-view-model __I__
  - can-view-nodelist __I__
  - can-view-parser __I__
  - can-view-scope __I__
  - can-view-target __I__
  - can-react-component __E__
  - react-view-model __E__
  - react-view-model/component __E__
  - can-stache-converters __E__

- Data Modeling
  - can-connect __C__
  - can-set __C__
  - can-connect-cloneable __E__
  - can-connect-feathers __E__
  - can-connect-ndjson __E__
  - can-connect-signalr __E__
  - can-fixture __E__
  - can-fixture-socket __E__
  - can-ndjson-stream __E__

- DOM Utilities
  - can-control __I__
  - can-dom-events __I__
  - can-event-dom-enter __I__
  - can-event-dom-radiochange __I__
  - can-jquery

- Routing
  - can-route __C__
  - can-route-pushstate __C__
  - can-param __I__
  - can-deparam __I__
  - can-stache/helpers/route

- Typed Data
  - can-construct __I__
  - can-construct-super __E__
  - can-cid __I__
  - can-namespace __I__
  - can-reflect __I__

- Polyfills
  - can-symbol __I__
  - can-vdom

- JS Utilities
  - can-ajax
  - ...

### Design

- If someone clicks directly a link like "can-map", the `API Docs` should only show modules in their collection.
- If multiple collections are shown at once, we should sort packages first by collection, then by name.
- Collection icons should be color coded to show how "safe" a collection's tools are.  

### Development 

- We will probably need some form of the sidebar's data (or a reduced docMap) provided to the client.
- I think we should implement this entirely in client-js first, then maybe make pages get generated with the sidebar.

## Tasks

- [ ] - Get a design
- [ ] - Change every package to have a "purpose" group and tag it with its collection.
- [ ] - Make the sidebar able to load at least the packages graph.  Possibly have the sidebar completely generated in the client.

## Concerns

- Having the collection toggle at the top might be annoying because you wont be able to see what's there.
- Do we hide groupings that aren't present?
- If we remove most helpers to another project #3426 how does that change things?

**Comment by matthewp** [comment]
:+1 I like it

**Comment by frank-dspeed** [comment]
https://github.com/donejs/donejs/issues/991 is related i think

**Comment by justinbmeyer** [comment]
Here's rough idea of what it might look like with icons and only core selected:

![image](https://user-images.githubusercontent.com/78602/30554412-0cea64aa-9c6a-11e7-9009-e0f3e6f4b121.png)

If someone toggled the buttons, in this case the infrastructure button, they would see the infrastructure packages, likely color-coded:

![image](https://user-images.githubusercontent.com/78602/30554443-269e330e-9c6a-11e7-9aa3-aa27f6f297cd.png)

I've colored the ecosystem icon a dirty orange/brown and the legacy icon red to indicate they these are not as "solid" as core and infrastructure.

Some concerns:

- Will people understand the icons?  We can make sure the API page explains them.
- If someone clicks directly a link like "can-map", should they see the `API Docs` section "only" showing the other legacy modules?

**Comment by justinbmeyer** [comment]
One way to deal with the concerns that:
- expand / collapse might be far from the area the user is viewing
- we might hide groupings without core parts

Is to add buttons to each purpose grouping:


<img width="320" alt="api_docs" src="https://user-images.githubusercontent.com/78602/30556178-92ef0fb0-9c6f-11e7-9be1-085ad0482c0a.png">

**Comment by adrifolio** [comment]
Here's another idea:

1.) Show the entire name of each toggle
<img width="274" alt="screen shot 2017-09-18 at 1 03 51 pm" src="https://user-images.githubusercontent.com/950214/30558656-ccc319ea-9c77-11e7-9fdd-0709f81d6d9b.png">

2.) Show colored squares next to each package according to group they belong to. In the example below, `ecosystem` and `legacy` have not been selected.

<img width="321" alt="screen shot 2017-09-18 at 1 02 34 pm" src="https://user-images.githubusercontent.com/950214/30558719-11211db2-9c78-11e7-8b6b-97f775792e19.png">

Note that the color for each group would needs more consideration, I just picked some random colors to show this concept.

**Comment by bmomberger-bitovi** [comment]
@adrifolio's idea looks very nice, but please remember colorblind users.  Perhaps icon bullets and an outline style for inactive?

**Comment by chasenlehara** [comment]
This is finished and has been deployed to https://canjs.com
